k8s.io/apiserver/pkg/server/filters.TestTimeoutHeaders flakes

### Which jobs are flaking?

`pull-kubernetes-unit`

### Which tests are flaking?

Unit test `TestTimeoutHeaders` in `k8s.io/apiserver/pkg/server/filters` package flakes
```
=== RUN   TestTimeoutHeaders
    timeout_test.go:247: got res.StatusCode 200; expected 503
--- FAIL: TestTimeoutHeaders (0.00s)
```

### Since when has it been flaking?

https://storage.googleapis.com/k8s-triage/index.html?date=2022-01-25&pr=1&job=unit&test=TestTimeoutHeaders

### Testgrid link

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107910/pull-kubernetes-unit/1490675490954940416

### Reason for failure (if possible)

there is a race condition between the go routine that executes the request handler and the timeout go routine that waits for the handler to finish


### Relevant SIG(s)

/sig api-machinery

--- a/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/filters/timeout.go
@@ -138,7 +138,7 @@ func (t *timeoutHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 			}()
 		}()
 
-		postTimeoutFn()
+		defer postTimeoutFn()
 		tw.timeout(err)
 	}
 }

--- a/staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/filters/timeout_test.go
@@ -219,6 +219,7 @@ func TestTimeoutHeaders(t *testing.T) {
 		})
 	}
 
+	postTimeoutCh := make(chan struct{})
 	ts := httptest.NewServer(
 		withDeadline(
 			WithTimeout(
@@ -226,13 +227,18 @@ func TestTimeoutHeaders(t *testing.T) {
 					h := w.Header()
 					// trigger the timeout
 					cancel()
-					// mutate response Headers
-					for j := 0; j < 1000; j++ {
-						h.Set("Test", "post")
+					// keep mutating response Headers until the request times out
+					for {
+						select {
+						case <-postTimeoutCh:
+							return
+						default:
+							h.Set("Test", "post")
+						}
 					}
 				}),
 				func(req *http.Request) (*http.Request, bool, func(), *apierrors.StatusError) {
-					return req, false, func() {}, apierrors.NewServerTimeout(schema.GroupResource{Group: "foo", Resource: "bar"}, "get", 0)
+					return req, false, func() { close(postTimeoutCh) }, apierrors.NewServerTimeout(schema.GroupResource{Group: "foo", Resource: "bar"}, "get", 0)
 				},
 			),
 		),

